SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,203 @@
+                entries.Add(Enumerable.Range(0, count).Select(i => GetColumnValue(stmt, i)).ToArray());
             }
+            var result = raw.sqlite3_finalize(stmt);
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to finalize SQL query, result={result}\n\t{errorMessage}");
+            }
+            return entries;
         }
 
-        private static StorageException ToStorageException(SQLiteException e)
+        public void CreateTable(string tableName, string[] columnNames, string[] columnTypes)
         {
-            return new StorageException($"SQLite errorCode={e.Result}", e);
+            var tableClause = string.Join(",", Enumerable.Range(0, columnNames.Length).Select(i => $"{columnNames[i]} {columnTypes[i]}"));
+            var result = ExecuteNonSelectionSqlQuery($"CREATE TABLE IF NOT EXISTS {tableName} ({tableClause});");
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to create table, result={result}\n\t{errorMessage}");
+            }
         }
 
-        public async Task<int> DeleteAsync<T>(Expression<Func<T, bool>> pred) where T : new()
+        public int Count(string tableName, string columnName, object value)
         {
-            try
+            var result = ExecuteSelectionSqlQuery($"SELECT COUNT(*) FROM {tableName} WHERE {columnName} = ?;", new[] { value });
+            var count = (long)(result.FirstOrDefault()?.FirstOrDefault() ?? 0L);
+            return (int)count;
+        }
+
+        public IList<object[]> Select(string tableName, string columnName, object value, string excludeColumnName, object[] excludeValues, int? limit = null)
+        {
+            var whereClause = $"{columnName} = ?";
+            var args = new List<object> { value };
+            if (excludeValues?.Length > 0)
             {
-                var numDeleted = 0;
-                var table = _dbConnection.Table<T>();
-                var entries = await table.Where(pred).ToListAsync().ConfigureAwait(false);
-                foreach (var entry in entries)
-                {
-                    numDeleted += await _dbConnection.DeleteAsync(entry).ConfigureAwait(false);
-                }
-                return numDeleted;
+                whereClause += $" AND {excludeColumnName} NOT IN ({BuildBindingMask(excludeValues.Length)})";
+                args.AddRange(excludeValues);
             }
-            catch (SQLiteException e)
+            var limitClause = limit != null ? $" LIMIT {limit}" : string.Empty;
+            var query = $"SELECT * FROM {tableName} WHERE {whereClause}{limitClause};";
+            return ExecuteSelectionSqlQuery(query, args);
+        }
+
+        public void Insert(string tableName, string[] columnNames, ICollection<object[]> values)
```

Comment:
Looks like the only place we use it is a string. We should change the insert code all the way from interface to here I think.

Reply:
Disagree, we can introduce other types (for priority for example) to match with other SDK, it will lead to revert of revert.
`string` in parameter here doesn't look like the right thing.

Reply:
Also, there are others methods here that requires this abstraction now. Let's keep it harmonized.

I don't think that "it isn't used at the moment" is the reason to drop already implemented correct abstraction.